README.md
analogsea
=======

master branch: [![Build Status](https://api.travis-ci.org/sckott/analogsea.png?branch=master)](https://travis-ci.org/sckott/analogsea)
[![Build status](https://ci.appveyor.com/api/projects/status/ll9lcqafuw338q0h/branch/master)](https://ci.appveyor.com/project/sckott/analogsea/branch/master)

v2 branch: [![Build Status](https://api.travis-ci.org/sckott/analogsea.png?branch=v2)](https://travis-ci.org/sckott/analogsea)
[![Build status](https://ci.appveyor.com/api/projects/status/ll9lcqafuw338q0h/branch/v2)](https://ci.appveyor.com/project/sckott/analogsea/branch/v2)

`analogsea` is a general purpose R interface to the [Digital Ocean](https://www.digitalocean.com/) API, AND includes scripts to install various R things:

* R
* RStudio Server
* RStudio Shiny Server
* OpenCPU

The Digital Ocean v2 API is fully implemented, but we're still working on the installation scripts...

## Digital Ocean API

+ [API v1 docs](https://developers.digitalocean.com/v1/)
+ [API v2 docs](https://developers.digitalocean.com/v2/)

### v1 and v2

The version on the `master` branch holds code that works with the Digital Ocean API v1.

The `v2` branch holds code in development for Digital Ocean API v2.

## Quick start

### Install

```coffee
devtools::install_github("sckott/analogsea", ref="v2")
library("analogsea")
```

### Authenticate

DigitalOcean uses OAuth 2. So we do the typical song and dance you may be used to on the web. Here's how to do it:

* Login to your Digital Ocean account. Go to _Apps & API_ page.
* In the _Developer Applications_ part of the page, press button for _Register new application_. Then enter app details:
  * _Application name:_ anything is fine
  * _Application homepage:_ anything is fine, this more used for web oauth flows.
  * _Application description:_ give a note to yourself what this app is.
  * _Application callback URL:_ use http://localhost:1410
* Once the app is registered, you can enter your appname, client id, and client secret as parameters to `do_oauth()`, or store them in your `.Rprofile` file as `do_appname`, `do_client_id`, and `do_client_secret`, respectively.

If `do_oauth()` runs succesfully, your token object is returned and cached, which will be called in other functions in this package.

### Get droplets

All droplets

```coffee
droplets()
```

```coffee
$meta
$meta$total
[1] 3


$droplet_ids
[1] 2435583 2435584 2435586

$droplets
$droplets$data
       id         name memory vcpus disk locked status           created_at region            image
1 2435583     feminist    512     1   20   TRUE    new 2014-08-23T22:10:21Z   sfo1 Ubuntu 14.04 x64
2 2435584 cellulomonas    512     1   20   TRUE    new 2014-08-23T22:10:24Z   sfo1 Ubuntu 14.04 x64
3 2435586 singlehanded    512     1   20   TRUE    new 2014-08-23T22:10:27Z   sfo1 Ubuntu 14.04 x64

$droplets$details
       id region_slug     region_name region_available                              region_sizes region_features image_id image_distribution       image_slug
1 2435583        sfo1 San Francisco 1             TRUE 2gb,4gb,8gb,1gb,16gb,32gb,48gb,512mb,64gb  virtio,backups  5141286             Ubuntu ubuntu-14-04-x64
2 2435584        sfo1 San Francisco 1             TRUE 2gb,4gb,8gb,1gb,16gb,32gb,48gb,512mb,64gb  virtio,backups  5141286             Ubuntu ubuntu-14-04-x64
3 2435586        sfo1 San Francisco 1             TRUE 2gb,4gb,8gb,1gb,16gb,32gb,48gb,512mb,64gb  virtio,backups  5141286             Ubuntu ubuntu-14-04-x64
  image_public                           image_regions     image_created_at
1         TRUE nyc1,ams1,sfo1,nyc2,ams2,sgp1,lon1,nyc3 2014-07-23T17:08:52Z
2         TRUE nyc1,ams1,sfo1,nyc2,ams2,sgp1,lon1,nyc3 2014-07-23T17:08:52Z
3         TRUE nyc1,ams1,sfo1,nyc2,ams2,sgp1,lon1,nyc3 2014-07-23T17:08:52Z
                                                                                                                                                                                                                  image_action_ids
1 29202525,29203474,29204856,29204858,29204860,29204863,29204865,29204868,29209875,29209878,29209902,29209919,29209926,29209929,29209954,29209962,29209964,29209966,29209997,29209999,29210003,29210022,29210030,29210127,29210135
2 29202525,29203474,29204856,29204858,29204860,29204863,29204865,29204868,29209875,29209878,29209902,29209919,29209926,29209929,29209954,29209962,29209964,29209966,29209997,29209999,29210003,29210022,29210030,29210127,29210135
3 29202525,29203474,29204856,29204858,29204860,29204863,29204865,29204868,29209875,29209878,29209902,29209919,29209926,29209929,29209954,29209962,29209964,29209966,29209997,29209999,29210003,29210022,29210030,29210127,29210135
  size_slug size_transfer size_price_monthly size_price_hourly networks_.id networks_ip_address networks_netmask networks_gateway networks_type kernel_id
1     512mb             1                  5           0.00744           v4     107.170.229.251    255.255.255.0    107.170.229.1        public      1682
2     512mb             1                  5           0.00744           v4     198.199.113.180    255.255.255.0    198.199.113.1        public      1682
3     512mb             1                  5           0.00744           v4     192.241.235.225    255.255.255.0    192.241.235.1        public      1682
                                   kernel_name    kernel_version backup_ids snapshot_ids action_ids
1 * Ubuntu 14.04 x64 vmlinuz-3.13.0-32-generic 3.13.0-32-generic         NA           NA   31285248
2 * Ubuntu 14.04 x64 vmlinuz-3.13.0-32-generic 3.13.0-32-generic         NA           NA   31285250
3 * Ubuntu 14.04 x64 vmlinuz-3.13.0-32-generic 3.13.0-32-generic         NA           NA   31285256


$actions
$actions$id
NULL


$links
NULL
```

A single droplet. Pass in a list with details like above or a single droplet id.

```coffee
droplets(2435586)
```

```coffee
$meta
NULL

$droplet_ids
[1] 2435586

$droplets
$droplets$data
       id         name memory vcpus disk locked status           created_at region            image
1 2435586 singlehanded    512     1   20   TRUE    new 2014-08-23T22:10:27Z   sfo1 Ubuntu 14.04 x64

$droplets$details
       id region_slug     region_name region_available                              region_sizes region_features image_id image_distribution       image_slug
1 2435586        sfo1 San Francisco 1             TRUE 2gb,4gb,8gb,1gb,16gb,32gb,48gb,512mb,64gb  virtio,backups  5141286             Ubuntu ubuntu-14-04-x64
  image_public                           image_regions     image_created_at
1         TRUE nyc1,ams1,sfo1,nyc2,ams2,sgp1,lon1,nyc3 2014-07-23T17:08:52Z
                                                                                                                                                                                                                  image_action_ids
1 29202525,29203474,29204856,29204858,29204860,29204863,29204865,29204868,29209875,29209878,29209902,29209919,29209926,29209929,29209954,29209962,29209964,29209966,29209997,29209999,29210003,29210022,29210030,29210127,29210135
  size_slug size_transfer size_price_monthly size_price_hourly networks_.id networks_ip_address networks_netmask networks_gateway networks_type kernel_id
1     512mb             1                  5           0.00744           v4     192.241.235.225    255.255.255.0    192.241.235.1        public      1682
                                   kernel_name    kernel_version backup_ids snapshot_ids action_ids
1 * Ubuntu 14.04 x64 vmlinuz-3.13.0-32-generic 3.13.0-32-generic         NA           NA   31285256


$actions
$actions$id
[1] 31285256


$links
NULL
```

### Spin up a new droplet

To make this as dead simple as possible, you just use one function, without any parameters.

```coffee
droplets_new()
```

```coffee
$droplet
$droplet$id
[1] 2435596

$droplet$name
[1] "copable"

$droplet$memory
[1] 512

$droplet$vcpus
[1] 1

$droplet$disk
[1] 20

$droplet$region
$droplet$region$slug
[1] "sfo1"

$droplet$region$name
[1] "San Francisco 1"

$droplet$region$sizes
[1] "2gb"   "4gb"   "8gb"   "1gb"   "16gb"  "32gb"  "48gb"  "512mb" "64gb"

$droplet$region$available
[1] TRUE

$droplet$region$features
[1] "virtio"  "backups"


...cutoff
```


Alternatively, you can of course pass in lots of options for name of the droplet, RAM size, disk size, ssh keys, etc.

I strongly suggest using SSH keys.

```coffee
droplets_new(name="foo", size = '512mb', image = 'ubuntu-14-04-x64', region = 'sfo1', ssh_keys = 89103)
```


#### Delete a droplet


```coffee
droplets_delete(1707487)
```

```coffee
success: (204) No Content
```

#### Regions, and their details

```coffee
regions()
```

```coffee
$regions
  slug            name                                             sizes available                                  features
1 nyc1      New York 1                                                       FALSE                           virtio, backups
2 ams1     Amsterdam 1                                                       FALSE                           virtio, backups
3 sfo1 San Francisco 1 2gb, 4gb, 8gb, 1gb, 16gb, 32gb, 48gb, 512mb, 64gb      TRUE                           virtio, backups
4 nyc2      New York 2 1gb, 2gb, 4gb, 8gb, 32gb, 64gb, 512mb, 16gb, 48gb      TRUE       virtio, private_networking, backups
5 ams2     Amsterdam 2 512mb, 1gb, 2gb, 4gb, 8gb, 32gb, 48gb, 64gb, 16gb      TRUE       virtio, private_networking, backups
6 sgp1     Singapore 1 1gb, 512mb, 2gb, 4gb, 8gb, 16gb, 32gb, 48gb, 64gb      TRUE virtio, private_networking, backups, ipv6
7 lon1        London 1 512mb, 1gb, 2gb, 4gb, 8gb, 16gb, 32gb, 48gb, 64gb      TRUE virtio, private_networking, backups, ipv6
8 nyc3      New York 3 512mb, 1gb, 2gb, 4gb, 8gb, 16gb, 32gb, 48gb, 64gb      TRUE virtio, private_networking, backups, ipv6

$meta
$meta$total
[1] 8
```

### RAM and disk sizes, and their details

```coffee
sizes()
```

```coffee
$sizes
   slug memory vcpus disk transfer price_monthly price_hourly                                        regions
1 512mb    512     1   20        1             5      0.00744 nyc1, sgp1, ams1, ams2, sfo1, nyc2, lon1, nyc3
2   1gb   1024     1   30        2            10      0.01488 nyc1, nyc2, sgp1, ams1, sfo1, ams2, lon1, nyc3
3   2gb   2048     2   40        3            20      0.02976 nyc1, nyc2, sfo1, ams1, sgp1, ams2, lon1, nyc3
4   4gb   4096     2   60        4            40      0.05952 nyc2, sfo1, ams1, sgp1, ams2, nyc1, lon1, nyc3
5   8gb   8192     4   80        5            80      0.11905 nyc2, sfo1, sgp1, ams1, ams2, nyc1, lon1, nyc3
6  16gb  16384     8  160        6           160      0.23810       sgp1, nyc1, sfo1, lon1, ams2, nyc3, nyc2
7  32gb  32768    12  320        7           320      0.47619       nyc2, sgp1, ams2, nyc1, sfo1, lon1, nyc3
8  48gb  49152    16  480        8           480      0.71429       sgp1, ams2, sfo1, nyc1, lon1, nyc3, nyc2
9  64gb  65536    20  640        9           640      0.95238       sgp1, ams2, nyc1, nyc2, sfo1, lon1, nyc3

$meta
$meta$total
[1] 9
```

### Chaining

Most of the `droplets_*` functions are meant to be chained together using the `%>%` operator.

For example, you can spin up a new droplet, then list your droplets, then

```r
droplets_new()
# output not shown
```

```r
droplets() %>%
  droplets_rename(name='mycoolnewname')
# output not shown
```

List the actions taken on a particular droplet

```r
droplets() %>%
  droplets_actions
```

```r
       id    status   type           started_at         completed_at resource_id resource_type region
1 31398316 completed rename 2014-08-25T21:35:26Z 2014-08-25T21:35:36Z     2451351       droplet   sfo1
2 31398261 completed create 2014-08-25T21:33:41Z 2014-08-25T21:34:31Z     2451351       droplet   sfo1
```

Notice that right now chaining only takes the first droplet from the `droplets()` function. So if you have 10 droplets, if you run `droplets()`, the first one will be passed to the next method if chained.

Let me know if you have a better idea for how to handle this.

Instead of chaining for `droplets_*` functions you can pass a droplet id itself as the first argument to another `droplets()` function, so you can easily loop across a bunch of droplet ids if needed.


### Install RStudio Server


__BEWARE:__ I'm working on updating these scripts so that it all actually works...hold on a bit...

You of course have to have an active droplet to do this. Once you have one, pass in a droplet id.

```coffee
do_install(1746479, what='rstudio', usr='hey', pwd='there')
```

This will install R, RStudio Server and it's dependencies. It will automatically pop open the RStudio server instance in your default browser.


NEWS
analogsea 0.1.0
===============

NEW FEATURES

* nothing yet...


R/zzz.R
#' GET request and output result or errors
#'
#' @import httr jsonlite assertthat XML
#' @export
#' @keywords internal
#' @param what What to return, parsed or raw
#' @param path Path to append to the end of the base Digital Ocean API URL
#' @param query Arguments to GET
#' @param parse To parse result to data.frame or to list
#' @param config Options passed on to httr::GET. Must be named, see examples.
#' @return Some combination of warnings and httr response object, data.frame, or list

do_GET <- function(what, path, query = NULL, parse=FALSE, config=NULL) {
  url <- file.path("https://api.digitalocean.com/v2", path)
  auth <- do_oauth()
  tt <- GET(url, query = query, config = c(token = auth, config))
  if(tt$status_code > 202){
    if(tt$status_code > 202) stop(content(tt)$message, call. = FALSE)
    if(content(tt)$status == "ERROR") stop(content(tt)$message, call. = FALSE)
  }
  if(what=='parsed'){
    res <- content(tt, as = "text")
    jsonlite::fromJSON(res, parse)
  } else { tt }
#   auth <- add_headers(Authorization = sprintf('Bearer %s', au$token))
}

#' Digital Ocean POST request handler
#'
#' @export
#' @keywords internal
#' @param what What to return, parsed or raw
#' @param path Path to append to the end of the base Digital Ocean API URL
#' @param args Arguments to POST
#' @param parse To parse result to data.frame or to list
#' @param config Options passed on to httr::GET. Must be named, see examples.
#' @param encodejson (logical) Whether to set \code{encode='json'} in \code{httr::POST} call.
#' @return Some combination of warnings and httr response object, data.frame, or list

do_POST <- function(what, path, args, parse=FALSE, config=config, encodejson=FALSE) {
  url <- file.path("https://api.digitalocean.com/v2", path)
  auth <- do_oauth()
  if(encodejson)
    tt <- POST(url, config = c(token = auth, config), body=args, encode="json")
  else
    tt <- POST(url, config = c(token = auth, config), body=args)
  if(tt$status_code > 202){
    if(tt$status_code > 202) stop(content(tt)$message, call. = FALSE)
    if(content(tt)$status == "ERROR") stop(content(tt)$message, call. = FALSE)
  }
  if(what=='parsed'){
    res <- content(tt, as = "text")
    jsonlite::fromJSON(res, parse)
  } else { tt }
}

#' Digital Ocean PUT request handler
#'
#' @export
#' @keywords internal
#' @param what What to return, parsed or raw
#' @param path Path to append to the end of the base Digital Ocean API URL
#' @param args Arguments to POST
#' @param parse To parse result to data.frame or to list
#' @param config Options passed on to httr::GET. Must be named, see examples.
#' @return Some combination of warnings and httr response object, data.frame, or list

do_PUT <- function(what, path, args, parse=FALSE, config=config) {
  url <- file.path("https://api.digitalocean.com/v2", path)
  auth <- do_oauth()
  tt <- PUT(url, config = c(token = auth, config), body=args)
  if(tt$status_code > 202){
    if(tt$status_code > 202) stop(content(tt)$message)
    if(content(tt)$status == "ERROR") stop(content(tt)$message)
  }
  if(what=='parsed'){
    res <- content(tt, as = "text")
    jsonlite::fromJSON(res, parse)
  } else { tt }
}


#' Digital Ocean DELETE request handler
#'
#' @export
#' @keywords internal
#' @param path Path to append to the end of the base Digital Ocean API URL
#' @param config Options passed on to httr::GET. Must be named, see examples.
#' @return Some combination of warnings and httr response object, data.frame, or list

do_DELETE <- function(path, config=NULL) {
  url <- file.path("https://api.digitalocean.com/v2", path)
  auth <- do_oauth()
  tt <- DELETE(url, config = c(token = auth, config))
  if(tt$status_code > 204){
    if(tt$status_code > 204) stop(content(tt)$message)
    if(content(tt)$status == "ERROR") stop(content(tt)$message)
  }
  if(http_status(tt)$category=='success'){
    message(http_status(tt)$message)
    invisible(http_status(tt)$message)
  } else { stop('Something went wrong') }
}

#' Compact
#'
#' @param ... List input
#' @keywords internal
ct <- function (...) Filter(Negate(is.null), list(...))

mssg <- function(x, y) if(x) message(y)

writefile <- function(filename, installstring){
  installrfile = filename
  fileConn <- file(installrfile)
  writeLines(installstring, fileConn)
  close(fileConn)
}

cli_tools <- function(ip){
  tmp <- Sys.which(c("ssh","scp"))
  nf <- names(tmp[vapply(tmp, nchar, 1) == 0])
  if(length(nf) != 0)
    stop(sprintf("%s not found on your computer\nTry ssh'ing into the machine\n    (ssh root@%s)\n& manually installing things. See ?do_scripts for help", nf, ip))
}

check_droplet <- function(x){
  if(class(x) == "response"){
    message("httr response object detected, passing")
    NULL
  } else {
    if(!is.null(x)){

      if(is.list(x)){
        if(length(x$droplet_ids) > 1) message("More than 1 droplet, using first")
        retid <- x$droplet_ids[[1]]
      } else {
        retid <- as.numeric(as.character(x))
      }
      if(!is.numeric(retid)) stop("Could not detect a droplet id")

      # check actions, and wait if not 'completed' or 'errored' status
      if(!x$droplets$data$status[1] == 'active'){
        actiondat <- x['actions']
        if(is.na(actiondat)){ return( retid ) } else {
          if(!is.null(actiondat$actions$id)){
            actionid <- actiondat$actions$id
            if(length(actionid) > 1) actionid <- actionid[1]
            tocheck <- 0
            while(tocheck != 1){
              actioncheck <- actions(x = actionid)
              tocheck <- if(actioncheck$action$status %in% c('completed','errored')) 1 else 0
            }
            return( retid )
          } else { return( retid ) }
        }
      } else { return( retid) }
    } else { NULL }
  }
}

match_droplet <- function(x, id){
  if(length(x$droplet_ids) > 1){
    x$droplets$data <- x$droplets$data[ x$droplets$data$id %in% id, ]
    x$droplets$details <- x$droplets$details[ x$droplets$details$id %in% id, ]
  }
  x[ !names(x) %in% c('meta','actions') ]
}

parse_to_df <- function(tmp){
  if(length(tmp) == 1){
    tmp[[1]][vapply(tmp[[1]], is.null, logical(1))] <- NA
    data.frame(tmp[[1]], stringsAsFactors = FALSE)
  } else {
    do.call(rbind.fill, lapply(tmp[[1]], function(z){
      z[vapply(z, is.null, logical(1))] <- NA
      data.frame(z, stringsAsFactors = FALSE)
    }))
  }
}

#' Rate limit information for the authenticated user.
#'
#' @export
#' @param x input to print, a do_rate S3 object
#' @param ... Not used

do_rate_limit <- function(){
  url <- "https://api.digitalocean.com/v2/sizes"
  auth <- do_oauth()
  tt <- HEAD(url, config = list(token = auth))
  tmp <- as.numeric(tt$headers[c('ratelimit-limit','ratelimit-remaining','ratelimit-reset')])
  reset <- as.POSIXct(tmp[3], origin="1970-01-01")
  dat <- list(limit=tmp[1], remaining=tmp[2], reset=reset)
  class(dat) <- "do_rate"
  dat
}

#' @method print do_rate
#' @export
#' @rdname do_rate_limit
print.do_rate <- function(x, ...){
  cat("Rate limit:", x$limit, '\n')
  cat("Limit remaining:", x$remaining, '\n')
  cat("Reset time:", '\n')
  print(x$reset)
}

#' Set Digital Ocean options including ssh keys, etc.
#'
#' This function sets options and prints them so you know what options are set.
#'
#' @export
#' @param size (optional) A Digital Ocean size slug name, e.g. '1gb'. Saved in options as 'do_size'
#' @param image (optional) A Digital Ocean image name, e.g., 'ubuntu-14-04-x64'. Saved in
#' options as 'do_image'
#' @param region (optional) A Digital Ocean region name, e.g., 'nyc1'. Saved in options
#' as 'do_region'
#' @param ssh_keys (optional) One or more ssh key id numbers or fingerprints. Put many in a list
#' or vector. Saved in options as 'do_ssh_keys'
#' @param private_networking (optional) A logical, whether to use private networking or not.
#' Saved in options as 'do_private_networking'
#' @param backups (optional) A logical, whether to enable backups. Automated backups can only be
#' enabled when the Droplet is created. Saved in options as 'do_backups'
#' @param ipv6 (optional) A boolean indicating whether IPv6 is enabled on the Droplet. Saved in
#' options as 'do_ipv6'
#' @param unset (optional) A boolean. If TRUE, unsets options so as to use defaults in 
#' \code{droplets_new}. If FALSE (default) your options are used in \code{droplets_new}.
#'
#' @details
#' These options are read and used by \code{droplets_new}.
#'
#' You can only set one value for each of size, image, and region, but multiple
#' values for ssh_keys as you can use multiple ssh keys on one DO droplet.
#'
#' Keep in mind that there are defaults set for size, image, and region in \code{droplets_new}.
#'
#' @examples \dontrun{
#' do_options(ssh_keys=109789)
#' getOption('do_ssh_keys')
#' do_options(size="1gb")
#' do_options(size="1gb", image='ubuntu-14-04-x64', region='nyc1')
#' getOption('do_size')
#' getOption('do_image')
#' getOption('do_region')
#' }
do_options <- function(size=NULL, image=NULL, region=NULL, ssh_keys=NULL, private_networking=NULL,
  backups=NULL, ipv6=NULL, unset=FALSE)
{
  if(!unset){
    if(!is.null(size)) options(do_size = size)
    if(!is.null(image)) options(do_image = image)
    if(!is.null(region)) options(do_region = region)
    if(!is.null(ssh_keys)) options(do_ssh_keys = ssh_keys)
    if(!is.null(private_networking)) options(do_private_networking = private_networking)
    if(!is.null(backups)) options(do_backups = backups)
    if(!is.null(ipv6)) options(do_ipv6 = ipv6)
  } else {
    options(do_size = NULL); options(do_image = NULL); options(do_region = NULL)
    options(do_ssh_keys = NULL); options(do_private_networking = NULL)
    options(do_backups = NULL); options(do_ipv6 = NULL)
  }

  cat("Your analogsea default options for spinning up a new droplet:", "\n")
  cat("[size]", gopt('do_size', 'not set (Defaults to: 512mb)'), "\n")
  cat("[image]", gopt('do_image', 'not set (Defaults to: ubuntu-14-04-x64)'), "\n")
  cat("[region]", gopt('do_region', 'not set (Defaults to: sfo1)'), "\n")
  cat("[ssh keys]", gopt('do_ssh_keys'), "\n")
  cat("[private networking]", gopt('do_private_networking'), "\n")
  cat("[backups]", gopt('do_backups'), "\n")
  cat("[ipv6]", gopt('do_ipv6'))
}

gopt <- function(x, y='not set') getOption(x, y)
gopt2 <- function(c, a, b) if(!is.null(c)) eval(c) else getOption(a, b)

nn <- function(x, unbox=TRUE){
  z <- switch(deparse(substitute(x)), 
         name = eval(x),
         size = gopt2(x, 'do_size', '512mb'),
         image = gopt2(x, 'do_image', 'ubuntu-14-04-x64'),
         region = gopt2(x, 'do_region', 'sfo1'),
         ssh_keys = gopt2(x, 'do_ssh_keys', NULL),
         private_networking = gopt2(x, 'do_private_networking', NULL),
         backups = gopt2(x, 'do_backups', NULL),
         ipv6 = gopt2(x, 'do_ipv6', NULL)
  )
  if(is.null(z)) z else if(unbox) unbox(z) else z
}

foo <- function(x){
  deparse(substitute(x))
}
